app/controllers/spree/admin/products_controller.rb

```diff
@@ -76,10 +76,13 @@ def bulk_update
 
         product_set.collection.each { |p| authorize! :update, p }
 
-        if product_set.save
+        saved = product_set.save
+        errors = product_and_variants_errors(product_set)
+
+        if saved && errors.empty?
           redirect_to main_app.bulk_products_api_v0_products_path(bulk_index_query)
-        elsif product_set.errors.present?
-          render json: { errors: product_set.errors }, status: :bad_request
+        elsif errors.any?
+          render json: { errors: errors }, status: :bad_request
```

Comment:
Oh, I just realise that ProductSet has an `errors` method already. I guess that the only problem is that it includes only the product errors and not the variant errors. If we add the variant errors to the products then this code can stay the same. I show you what I mean in the next comment.

Reply:
Thanks!